tactics/tactics.ml

```diff
@@ -2249,14 +2249,15 @@ let one_constructor i lbind = constructor_tac false None i lbind
    Should be generalize in Constructor (Fun c : I -> tactic)
  *)
 
-let rec tclANY tac = function
-| [] -> Tacticals.New.tclZEROMSG (str "No applicable tactic.")
-| arg :: l ->
-  Tacticals.New.tclORD (tac arg) (fun () -> tclANY tac l)
-
 let any_constructor with_evars tacopt =
-  let t = match tacopt with None -> Proofview.tclUNIT () | Some t -> t in
-  let tac i = Tacticals.New.tclTHEN (constructor_tac with_evars None i NoBindings) t in
+  let one_constr =
+    let tac i = constructor_tac with_evars None i NoBindings in
+    match tacopt with
+    | None -> tac
+    | Some t -> fun i -> Tacticals.New.tclTHEN (tac i) t in
+  let rec any_constr n i () =
+    if Int.equal i n then one_constr i
+    else Tacticals.New.tclORD (one_constr i) (any_constr n (i + 1)) in
```

Comment:
You're building the whole tactic list eagerly on the stack here. It's not worse than before, but now that you are rolling a custom recursion, you may want to suspend the recursive call.

Reply:
The recursive call is suspended. Notice that this is the `tclORD` tactical, where `D` means suspended (or something close to it). Notice also that the `any_constr` function takes three arguments, but only two of them are passed to the recursive call, which thus is not a recursive call.

Reply:
Ah yes. Nevermind me. `D` is for « delayed ». I think I introduced this terminology…

All good then.